retis-events/src/file.rs

```diff
@@ -2,47 +2,113 @@
 
 use std::{
     fs::File,
-    io::{BufRead, BufReader},
+    io::{BufRead, BufReader, Seek},
     path::Path,
-    time::Duration,
 };
 
-use anyhow::{anyhow, Result};
+use anyhow::{anyhow, bail, Result};
 
-use super::{Event, EventResult};
+use super::{Event, EventSeries};
+
+// Type of file that is being processed.
+#[derive(Debug)]
+pub enum FileType {
```

Comment:
> I like the general direction of being able to store sorted series and use them later (print, python). There are a few opened questions though.
> 
> It makes `retis sort` to display the sorted result when an unsorted set of event is given; but that doesn't work for a sorted one. I'm not saying this is not logical, but this makes the logic a bit hard to follow for users:
> 
>     * Unsorted events -> `retis print` or `retis sort`
> 
>     * Sorted events -> `retis print`
> 

Would printing `File already sorted` help?

> 
> From an implementation point of view, I didn't see any pitfall but the result looks a bit convoluted. We're doing nearly the same thing with events and series but have long conditional blocks.
> 
> In Python the following is a bit unexpected IMO:
> 
> ```python
> >>> series = next(events)
> >>> ev = next(series)
> ```
> 

We could have helpers to explicitly iterate over series or event although in python typing is quite loose. In any case, maybe the source of confusion is the name of the global variable. If we replace it with `reader`, it would be better IMO.

> I don't have a clean solution in mind, but this goes in the same direction as a discussion we started about potentially merging events and series in a single (user fronting) representation.
> 
> I don't want this to be a blocker for this, but that would for sure make things cleaner from the start (given we can find a common representation).

I really don't know how would merging events and series would makes things clearer for the user. They will need to check if the (now common) type, `Event` is sorted or not, or has a "next-event" of some kind. That check would have to come on every event since they cannot assume the internals of one event based on the previous (plus sorted files would have single events).

For me that's more complicated than just having a check at the reader level and knowing what type it'll return (maybe with an explicit iterator helper).

Also, we would be loosing the ability to add series-specific helpers, currently it implements `len()` prototype which does not make sense for single events but we could add more such as a series comparator (did these two packets go though the same funcs?). If we add these helpers to single events users will have to deal with custom error codes or semantic differences if they are just single ones.

Maybe I'm just failing to see how the API would look like and how it'd be simpler.

Reply:
> > I like the general direction of being able to store sorted series and use them later (print, python). There are a few opened questions though.
> > It makes `retis sort` to display the sorted result when an unsorted set of event is given; but that doesn't work for a sorted one. I'm not saying this is not logical, but this makes the logic a bit hard to follow for users:
> > ```
> > * Unsorted events -> `retis print` or `retis sort`
> > 
> > * Sorted events -> `retis print`
> > ```
> 
> Would printing `File already sorted` help?

That's a good solution before (or if we can't) handle events and series in a similar way in our code.

> > From an implementation point of view, I didn't see any pitfall but the result looks a bit convoluted. We're doing nearly the same thing with events and series but have long conditional blocks.
> > In Python the following is a bit unexpected IMO:
> > ```python
> > >>> series = next(events)
> > >>> ev = next(series)
> > ```
> 
> We could have helpers to explicitly iterate over series or event although in python typing is quite loose. In any case, maybe the source of confusion is the name of the global variable. If we replace it with `reader`, it would be better IMO.

I liked the `events` name as it makes it quite explicit. But yes, with series that does work that well. This is an issue only in the interpreter though, maybe we can name it `events` or `series` based on context. Having a more generic name like `reader` is fine too, although I find `reader` a bit more linked to the underlying concept than what it contains. That's probably only personal taste though so not a strong opinion.

> > I don't have a clean solution in mind, but this goes in the same direction as a discussion we started about potentially merging events and series in a single (user fronting) representation.
> > I don't want this to be a blocker for this, but that would for sure make things cleaner from the start (given we can find a common representation).
> 
> I really don't know how would merging events and series would makes things clearer for the user. They will need to check if the (now common) type, `Event` is sorted or not, or has a "next-event" of some kind. That check would have to come on every event since they cannot assume the internals of one event based on the previous (plus sorted files would have single events).

I agree but that is also the case now in Python, the reader either return events or series. This makes using it very context dependent. There's `sorted()`, this is maybe the answer (although if one forget to call it right away, the script could fail).

> For me that's more complicated than just having a check at the reader level and knowing what type it'll return (maybe with an explicit iterator helper).
> 
> Also, we would be loosing the ability to add series-specific helpers, currently it implements `len()` prototype which does not make sense for single events but we could add more such as a series comparator (did these two packets go though the same funcs?). If we add these helpers to single events users will have to deal with custom error codes or semantic differences if they are just single ones.
> 
> Maybe I'm just failing to see how the API would look like and how it'd be simpler.

Those are all good points and I don't have a clear idea what the object nor the API could look like. I agree it's probably best to keep the `Event` and `Series` object different. There might be a way to improve their use though.

Reply:
> > > I like the general direction of being able to store sorted series and use them later (print, python). There are a few opened questions though.
> > > It makes `retis sort` to display the sorted result when an unsorted set of event is given; but that doesn't work for a sorted one. I'm not saying this is not logical, but this makes the logic a bit hard to follow for users:
> > > ```
> > > * Unsorted events -> `retis print` or `retis sort`
> > > 
> > > * Sorted events -> `retis print`
> > > ```
> > 
> > 
> > Would printing `File already sorted` help?
> 
> That's a good solution before (or if we can't) handle events and series in a similar way in our code.

Or maybe sorting a sorted input should just return the same thing?

Reply:
> > 
> > We could have helpers to explicitly iterate over series or event although in python typing is quite loose. In any case, maybe the source of confusion is the name of the global variable. If we replace it with `reader`, it would be better IMO.
> 
> I liked the `events` name as it makes it quite explicit. But yes, with series that does work that well. This is an issue only in the interpreter though, maybe we can name it `events` or `series` based on context. Having a more generic name like `reader` is fine too, although I find `reader` a bit more linked to the underlying concept than what it contains. That's probably only personal taste though so not a strong opinion.
> 
> > > I don't have a clean solution in mind, but this goes in the same direction as a discussion we started about potentially merging events and series in a single (user fronting) representation.
> > > I don't want this to be a blocker for this, but that would for sure make things cleaner from the start (given we can find a common representation).
> > 
> > 
> > I really don't know how would merging events and series would makes things clearer for the user. They will need to check if the (now common) type, `Event` is sorted or not, or has a "next-event" of some kind. That check would have to come on every event since they cannot assume the internals of one event based on the previous (plus sorted files would have single events).
> 
> I agree but that is also the case now in Python, the reader either return events or series. This makes using it very context dependent. There's `sorted()`, this is maybe the answer (although if one forget to call it right away, the script could fail).
> 

The difference is, with the reader you only need to check once. Then you call the right iterator, let's say `iter_series()` or `iter_events` and that's it. Moving those checks inside of the loop makes things complicated as you know need to consider both seriese and events being read from the same file. 

Having different iterators would make a script only supporting events not fail if called on a sorted file and vice-versa.